### ISSUE TITLE
missing unsqueeze(1) call in python notebook

### ISSUE THREAD

**Issue by owlshrimp** [brandontrabucco/nerf#2]
Hi,

Due to the issue in https://github.com/brandontrabucco/nerf/issues/1, only the example.ipynb example has the necessary data included to run (tiny_nerf_data.npz). However, the code in this notebook is missing an unsqueeze(1) call in the following line and crashes:

((pixels - batch['pixels']) ** 2).mean().backward()

I *think* the correct version is:

((pixels - batch['pixels'].unsqueeze(1)) ** 2).mean().backward()

since that's how it's done in train_conditional_nerf.py

One other note: it's quite convenient to copy the code in this notebook out into a separate *.py file, as that file plus tiny_nerf_data.npz yields a working example that doesn't need to be run on google colab or another notebook environment.

**Comment by owlshrimp** [comment]
my slightly hacked-up colab_example.py is as follows:

from nerf.model import NeRF
from nerf.dataset import PixelRayDataset
from tqdm import tqdm
import time
from datetime import timedelta
import os

import matplotlib.pyplot as plt
import pickle as pkl
import numpy as np
import cv2
import os
import io

import torch
import torch.utils.data as data
import torch.optim as optim
import torch.nn as nn
import torch.autograd as autograd


tiny_nerf_data = np.load('tiny_nerf_data.npz')
images = tiny_nerf_data['images']
poses = tiny_nerf_data['poses']

H, W = images.shape[1:3]
focal_length = float(tiny_nerf_data['focal'])


test_image = images[-1]
test_pose = poses[-1]

test_image = torch.FloatTensor([test_image]).cuda()
test_pose = torch.FloatTensor([test_pose]).cuda()

images = images[:-1]
poses = poses[:-1]

images = torch.FloatTensor(images).cuda()
poses = torch.FloatTensor(poses).cuda()


dataset = PixelRayDataset(images, poses, focal_length)
data_loader = data.DataLoader(dataset, batch_size=1024, shuffle=True)


nerf = NeRF(normalize_position=6.0).cuda()
#nerf = NeRF(normalize_position=6.0, num_stages=4).cuda()


nerf_optimizer = optim.Adam(nerf.parameters(), lr=0.0001)


psnrs = []
iternums = []

num_epochs = 200
output_dir = "images"

print("number of epochs: {}".format(num_epochs))
num_iters = len(data_loader)
print("iterations per epochs: {}".format(num_iters))
total_iters = num_epochs*num_iters
print("the total number of iterations will be: {}".format(total_iters))

if not os.path.exists(os.path.join(".",output_dir)):
    os.makedirs(os.path.join(".",output_dir))

start = time.time()
iteration = -1
for epoch in range(num_epochs):
    with tqdm(total=num_iters) as pbar:
        for batch in data_loader:
            
            iteration += 1
            pbar.update(1)
            
            #print("rays o shape: {}".format(batch['rays_o'].shape))
            #print("rays d shape: {}".format(batch['rays_d'].shape))
            
            #print("num stages: {}".format(nerf.num_stages))
            
            pixels = nerf.render_rays(
                batch['rays_o'],
                batch['rays_d'],  
                2.0, 
                6.0, 
                64, 
                randomly_sample=True, 
                density_noise_std=1.0)

            nerf_optimizer.zero_grad()

            #print("pixels shape: {}".format(pixels.shape))
            #print("batch pixels shape: {}".format(batch['pixels'].shape))
            #print("batch pixels unsqueeze shape: {}".format(batch['pixels'].unsqueeze(1).shape))
            
            #((pixels - batch['pixels']) ** 2).mean().backward()
            ((pixels - batch['pixels'].unsqueeze(1)) ** 2).mean().backward()

            nerf_optimizer.step()
            
            #if iteration % 1000 == 0:
            if iteration % num_iters == 0: # save once per epoch (currently 1026 iters)
                elapsed = (time.time() - start)
                
                with torch.no_grad():
                    
                    test_render = nerf.render_image(
                        test_pose[..., :3,  3], 
                        test_pose[..., :3, :3],  
                        H, 
                        W, 
                        focal_length, 
                        2.0, 
                        6.0, 
                        64)
            
                    psnr = -10.0 * torch.log(((
                        test_render - 
                        test_image) ** 2).mean()) / 2.30258509299
                
                    psnrs.append(psnr.cpu().detach().numpy())
                    iternums.append(iteration)
            
                plt.figure(figsize=(15,4))
                
                plt.subplot(131)
                plt.title(f'Iteration: {iteration}')
                plt.imshow(test_render[0].detach().cpu().numpy())
                
                plt.subplot(132)
                plt.title('Ground Truth')
                plt.imshow(test_image[0].detach().cpu().numpy())
                
                plt.subplot(133)
                plt.plot(iternums, psnrs)
                plt.title('PSNR')
                
                #plt.show()
                print("epoch: {}".format(epoch))
                name = "out_{:07d}.png".format(iteration)
                print("saving {}".format(name))
                plt.savefig(os.path.join(".",output_dir,name))
                #plt.close()
                print("elapsed time: {}".format(str(timedelta(seconds=elapsed))))
#plt.show()

**Comment by brandontrabucco** [comment]
Hello Matt, this notebook is a bit out of date because I updated the nerf model to use the hierarchical volume sampling method described in the NeRF paper, but I wrote the notebook before that. I will push an update to the notebook soon which will fix this shape issue with image generation.

**Comment by owlshrimp** [comment]
Is the use of .unsqueeze(1) like that at least correct? it seems to produce reasonably correct results.

**Comment by brandontrabucco** [comment]
Yep!

It is correct, there is an additional dimension at position 1 for the number of stages of volume sampling.
